ambry-store/src/main/java/com.github.ambry.store/PersistentIndex.java

```diff
@@ -1055,57 +1055,20 @@ BlobReadOptions getBlobReadInfo(StoreKey id, EnumSet<StoreGetOptions> getOptions
-            metrics.putEntryDeletedInfoMismatchCount.inc();
-          }
-          if (putValue.getSize() != deletedBlobInfo.getSize()) {
-            logger.error("Size in PUT index entry {} is different from that in the PUT record {} for ID {}",
-                putValue.getSize(), deletedBlobInfo.getSize(), key);
-            metrics.putEntryDeletedInfoMismatchCount.inc();
-          }
-        }
-        Offset offset = new Offset(logSegmentName, value.getOriginalMessageOffset());
-        // use the expiration time from the original value because it may have been updated
-        readOptions = new BlobReadOptions(log, offset,
-            new MessageInfo(deletedBlobInfo.getStoreKey(), deletedBlobInfo.getSize(), true,
-                value.isFlagSet(IndexValue.Flags.Ttl_Update_Index), value.getExpiresAtMs(),
-                deletedBlobInfo.getAccountId(), deletedBlobInfo.getContainerId(),
-                deletedBlobInfo.getOperationTimeMs()));
-      } else if (putValue != null) {
-        // PUT record in a different log segment.
-        // use the expiration time from the original value because it may have been updated
-        readOptions = new BlobReadOptions(log, putValue.getOffset(),
-            new MessageInfo(key, putValue.getSize(), true, value.isFlagSet(IndexValue.Flags.Ttl_Update_Index),
-                value.getExpiresAtMs(), putValue.getAccountId(), putValue.getContainerId(),
-                putValue.getOperationTimeInMs()));
-      } else {
-        // PUT record no longer available.
-        throw new StoreException("Did not find PUT index entry for key [" + key
-            + "] and the the original offset in value of the DELETE entry was [" + value.getOriginalMessageOffset()
-            + "]", StoreErrorCodes.ID_Deleted);
-      }
-    } catch (IOException e) {
-      StoreErrorCodes errorCode = StoreException.resolveErrorCode(e);
-      throw new StoreException(errorCode.toString() + " when reading delete blob info from the log " + dataDir, e,
-          errorCode);
+    IndexValue putValue =
+        findKey(key, new FileSpan(getStartOffset(indexSegments), value.getOffset()), EnumSet.of(IndexEntryType.PUT),
+            indexSegments);
+    if (putValue != null) {
+      // use the expiration time from the original value because it may have been updated
+      readOptions = new BlobReadOptions(log, putValue.getOffset(),
+          new MessageInfo(key, putValue.getSize(), true, value.isFlagSet(IndexValue.Flags.Ttl_Update_Index),
+              value.getExpiresAtMs(), putValue.getAccountId(), putValue.getContainerId(),
+              putValue.getOperationTimeInMs()));
+    } else {
+      // PUT record no longer available.
+      throw new StoreException("Did not find PUT index entry for key [" + key
+          + "] and the the original offset in value of the DELETE entry was [" + value.getOriginalMessageOffset() + "]",
+          StoreErrorCodes.ID_Deleted);
```

Comment:
I wonder why the error code is ID_Deleted rather than ID_Not_Found?

Reply:
I think it can be further detailed. It means the original Put Record is already compacted, or this node only replicates the Delete record from it's peer, which are different error code and ambry frontend would react differently. But this is what we have before so I will just keep as it's.